src/UrlGenerator.php

```diff
@@ -0,0 +1,338 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Yiisoft\Router\FastRoute;
+
+use Nyholm\Psr7\Uri;
+use Yiisoft\Router\Group;
+use Yiisoft\Router\Route;
+use Yiisoft\Router\RouteCollectorInterface;
+use Yiisoft\Router\RouteNotFoundException;
+use Yiisoft\Router\UrlMatcherInterface;
+use Yiisoft\Router\UrlGeneratorInterface;
+use FastRoute\RouteParser;
+
+use function array_key_exists;
+use function array_keys;
+use function implode;
+use function is_string;
+use function preg_match;
+
+final class UrlGenerator implements UrlGeneratorInterface
+{
+    /** @var string */
+    private string $uriPrefix = '';
+
+    /**
+     * All attached routes as Route instances
+     *
+     * @var Route[]
+     */
+    private array $routes = [];
+
+    /**
+     * @var UrlMatcherInterface $matcher
+     */
+    private UrlMatcherInterface $matcher;
+
+    /**
+     * @var RouteCollectorInterface $collector
+     */
+    private RouteCollectorInterface $collector;
+
+    /**
+     * @var RouteParser
+     */
+    private RouteParser $routeParser;
+
+    /**
+     * Constructor
+     *
+     * @param UrlMatcherInterface $matcher url matcher
+     * @param RouteCollectorInterface $collector route collector
+     * @param RouteParser|null $parser
+     */
+    public function __construct(
+        UrlMatcherInterface $matcher,
+        RouteCollectorInterface $collector,
+        RouteParser $parser = null
+    ) {
+        $this->matcher = $matcher;
+        $this->collector = $collector;
+        $this->routeParser = $parser ?? new RouteParser\Std();
```

Comment:
No. Use external dependency only. Default value is unneccessary

Reply:
> No. Use external dependency only. Default value is unneccessary

This time you're wrong. There is no need to use a factory for UrlGenerator, but you still can create your own factory and inject your own custom parser if it is necessary.

Reply:
In which cases we won't have route parser passed as an argument?

Reply:
> In which cases we won't have route parser passed as an argument?

When we get it from the container.

Reply:
You mean if `RouteParser` is not defined in container?

Reply:
> You mean if `RouteParser` is not defined in container?
Yes. We don't define it for `UrlGenerator` now because we build the router by the factory. In this way, we don't need to create a factory or set the default parser in the container.